src/mbgl/layout/symbol_projection.cpp

```diff
@@ -325,19 +373,31 @@ namespace mbgl {
         const mat4 glCoordMatrix = getGlCoordMatrix(posMatrix, pitchWithMap, rotateWithMap, state, pixelsToTileUnits);
         
         dynamicVertexArray.clear();
+        
+        bool useVertical = false;
 
         for (auto& placedSymbol : placedSymbols) {
+            // Don't do calculations for vertical glyphs unless the previous symbol was horizontal
+            // and we determined that vertical glyphs were necessary.
+            // Also don't do calculations for symbols that are collided and fully faded out
+            if (placedSymbol.hidden || (placedSymbol.writingModes == WritingModeType::Vertical && !useVertical)) {
+                hideGlyphs(placedSymbol.glyphOffsets.size(), dynamicVertexArray);
+                continue;
+            }
+            // Awkward... but we're counting on the paired "vertical" symbol coming immediately after its horizontal counterpart
+            useVertical = false;
```

Comment:
I think it would make sense to fix this as part of a larger refactoring -- I think `PlacedSymbol` probably doesn't need to exist as a separate class, and in this loop we could be iterating over `SymbolInstance` (or something like it), which would contain the paired horizontal/vertical glyphs in one item. I'd like to defer that to after this PR if possible, though.

Reply:
yep, we should get rid of `PlacedSymbol` or at least nest it within a `SymbolInstance` instead of a separate array. I originally added `PlacedSymbol` in -js to separate the data needed for reproduction into a separate array so that only it would be transferred to the main thread. Now that `SymbolInstance`s are also being transferred there's no reason for this.

> I'd like to defer that to after this PR if possible, though.

I agree